source/common/http/header_utility.h

```diff
@@ -65,30 +67,195 @@ class HeaderUtility {
                                                              const Http::LowerCaseString& key,
                                                              absl::string_view separator = ",");
 
-  // A HeaderData specifies one of exact value or regex or range element
-  // to match in a request's header, specified in the header_match_type_ member.
-  // It is the runtime equivalent of the HeaderMatchSpecifier proto in RDS API.
-  struct HeaderData : public HeaderMatcher {
-    HeaderData(const envoy::config::route::v3::HeaderMatcher& config,
-               Server::Configuration::CommonFactoryContext& factory_context);
+  // A shared behavior for HeaderDataMatcher types (exact, regex, etc.) that
+  // correspond to the HeaderMatchSpecifier proto in RDS API.
+  class HeaderDataBaseImpl : public HeaderMatcher {
+  public:
+    HeaderDataBaseImpl(const envoy::config::route::v3::HeaderMatcher& config)
+        : name_(config.name()), invert_match_(config.invert_match()),
+          treat_missing_as_empty_(config.treat_missing_header_as_empty()) {}
+
+    // HeaderMatcher
+    bool matchesHeaders(const HeaderMap& request_headers) const override {
+      const auto header_value = getAllOfHeaderAsString(request_headers, name_);
+      // If treat_missing_as_empty_ is false and there is no header value, most
+      // matchers (other than HeaderDataPresentMatch) will just return false here.
+      if (!treat_missing_as_empty_ && !header_value.result().has_value()) {
+        return false;
+      }
+
+      // If the header does not have value and the result is not returned in the
+      // code above, it means treat_missing_as_empty_ is set to true and we should
+      // treat the header value as empty.
+      const auto value = header_value.result().has_value() ? header_value.result().value() : "";
```

Comment:
Is the type of this `auto` a reference or a value? I *think* it's a value due to the `""`. Could you avoid a copy here if you use the `EMPTY_STRING` value instead of constructing a string here? Or maybe make it an `absl::string_view`?

Reply:
Done, thanks!